SpockOpaleye/src/ApiBase.hs

```diff
@@ -0,0 +1,157 @@
+removeRawDbRows table matchFunc = do
+  conn <- getConnection
+  liftIO $ runDelete conn table matchFunc 
+
+createDbRows :: (Show columnsW, D.Default QueryRunner columnsR haskells) 
+    =>  Table columnsW columnsR -> [columnsW] -> AppM [haskells]
+createDbRows table pgrows = do
+  auditLog $ "Create : " ++ (show pgrows)
+  conn <- getConnection
+  liftIO $ runInsertManyReturning conn table pgrows (\x -> x)
+
+updateDbRow :: (Show columnsW, HasId columnsR (Column PGInt4)) => Table columnsW columnsR -> Column PGInt4 -> columnsW -> AppM columnsW
+updateDbRow table row_id item = do
+  auditLog $ "Update :" ++ (show item)
+  conn <- getConnection
+  _ <- liftIO $ runUpdate conn table (\_ -> item) (matchFunc row_id) 
+  return item
+  where
+    matchFunc :: (HasId cmR (Column PGInt4)) => (Column PGInt4 -> cmR -> Column PGBool)
+    matchFunc itId item' = (item' ^. id) .== itId
+  
+createRow ::(
+    Show incoming,
+    Show columnsW,
+    HasCreatedat columnsW (Maybe (Column PGTimestamptz)),
+    HasUpdatedat columnsW (Column PGTimestamptz),
+    D.Default Constant incoming columnsW, D.Default QueryRunner returned row)
+    => Table columnsW returned -> incoming -> AppM row
+createRow table item = do
+  auditLog $ "Create : " ++ (show item)
+  currentTime <- liftIO $ fmap pgUTCTime getCurrentTime
+  let itemPg = (constant item) & createdat .~ (Just currentTime) & updatedat .~ (currentTime)
+  fmap (head) $ createDbRows table [itemPg] 
+
+updateRow :: (
+    Show columnsW
+    , Show haskells
+    , HasUpdatedat haskells UTCTime
+    , D.Default Constant haskells columnsW
+    , D.Default Constant itemId (Column PGInt4)
+    , HasId haskells itemId
+    , HasId columnsR (Column PGInt4)
+    )
+    => Table columnsW columnsR -> haskells -> AppM haskells
+updateRow table item = do
+  auditLog $ "Update : " ++ (show item)
+  let itId = item ^. id
+  currentTime <- liftIO getCurrentTime
+  let updatedItem = (putUpdatedTimestamp currentTime) item
+  _ <- updateDbRow table (constant itId) (constant updatedItem) 
+  return updatedItem
+  where
+    putUpdatedTimestamp :: (HasUpdatedat item (UTCTime)) => UTCTime -> item -> item
+    putUpdatedTimestamp timestamp  = updatedat .~ timestamp
+
+updateAuditableRow :: (
+    Show columnsW
+    , HasUpdatedat (Auditable haskells) UTCTime
+    , D.Default Constant haskells columnsW
+    , D.Default Constant itemId (Column PGInt4)
+    , HasId (Auditable haskells) itemId
+    , HasId columnsR (Column PGInt4)
+    )
+    => Table columnsW columnsR -> Auditable haskells -> AppM (Auditable haskells)
+updateAuditableRow table audti = do
+  --auditLog $ "Update : " ++ (show item)
+  let itId = audti ^. id
+  currentTime <- liftIO getCurrentTime
+  let updatedItem = (putUpdatedTimestamp currentTime) audti
+  let Auditable { _data = item, _log = _log} = updatedItem
+  _ <- updateDbRow table (constant itId) (constant item) 
+  insertIntoLog table itId "" _log
+  return audti
```

Comment:
is the in-memory audit log being blanked out after a successful update or insert?